### DISCUSSION TITLE
A few questions about Gif decoding

### DISCUSSION THREAD

**Discussion by lanyusan** [libvips/libvips#3158] [Help wanted]
Hi, 

As per the other issue earlier, I was very happy with libvips/net-vips in encoding gif generated from our in-house generative art code base.

Now I am thinking of migrating our decoding code to libvips/net-vips as well.  I have a few questions.

1. I understand libvips/net-vips decode gif into a single vipsimage object with a page height parameter that could be used to extract each frame. Is there a function/method to extract each frames or should I do some math to manually extract frames?
2. How to read each frame's delay and disposal meta data from vipsimage loaded from gifload?
3. Does this lib's gif loader do 'coalesce' , which is a huge problem for many gif decoders. Our code would need to get fully coalesced frames. Currently we use imagemegick for decoding, which has very high quality 'coalese' :

Below is from imagegaick's doc:

https://www.imagemagick.org/Usage/anim_basics/#coalesce

```
The "-coalesce" operation basically converts an image into exactly what the animation should look like after the previous frame has been correctly and the next sub-frame overlaid.

That is, instead of an animation sequence where each frame only represents the overlaid changes to the previous 'disposed' frame.
 This operator creates a complete view of the animation at each point, a bit like a true film strip, rather than an animation sequence. 
Such a sequence, known as a [Coalesced Animation]s much easier to study, edit, modify and re-optimize.
```

Many thanks.

**Comment by jcupitt** [comment]
Hello again,

1. You can use `crop` to extract frames if you need per-frame processing. You can often just process the entire strip, of course. Use `arrayjoin` to reassemble an array of frames into a strip.
2. There are metadata items for this, eg. `image.get("delay")`. Try `vipsheader -a image-filename` to see all metadata items for an image. You can't get disposal though.
3. The output is always a fully decoded (or coalesced, to use imagemagick's term) RGB or RGBA image.

**Reply by lanyusan** [reply]
Thanks for the quick response. If fully decoded then disposal isn't necessary.
